cmd/worker_pool.go

```diff
@@ -0,0 +1,74 @@
+package cmd
+
+import bosherr "github.com/cloudfoundry/bosh-utils/errors"
+
+type WorkerPool struct {
+	WorkerCount int
+}
+
+// Runs the given set of tasks in parallel using the configured number of worker go routines
+// Will stop adding new tasks if a task throws an error, but will wait for in-flight tasks to finish
+func (w WorkerPool) ParallelDo(tasks ...func() (interface{}, error)) ([]interface{}, error) {
```

Comment:
What type of error checking are you looking for? Usages of the cloudfoundry/workpool library are usually doing their own error management through channels and such if they need that. In my experience error checking and tolerance is pretty variable when doing this type of background worker processing, which is why we chose to not try to generalize it.

Your implementation seems a little more advanced since it is grouping return values and errors together for you. Which is fine, but you could use the workpool throttler like this:

```go
errChan := make(chan err)
resultChan := make(chan interface{})
task := func() {
  // do work...
  if err != nil {
    errChan <- err
  } else {
    resultChan <- result
  }
})
// build up more tasks....

throttler, err = workpool.NewThrottler(maxWorkers, tasks)
if err != nil { log.Fatal(err) }

for {
  select {
    case err := <-errChan:
      // do something with err
    case result := <-resultChan:
      // do something with resultChan
    default:
  }
}

throttler.Work()
close(errChan)
close(resultChan)
```

Not polished code and obviously wrong in some ways, but the general idea is there. And you can chose what you want to do, maybe you don't have results, maybe you don't have errors. It's up to you.